src/crypto/CrossSigning.js

```diff
@@ -101,12 +105,92 @@ export class CrossSigningInfo extends EventEmitter {
         };
     }
 
-    hasKeys() {
-        return Object.keys(this.keys).length > 0;
+    /**
+     * Check whether the private keys exist in secret storage.
+     *
+     * @param {SecretStorage} secretStorage The secret store using account data
+     * @returns {boolean} Whether all private keys were found in storage
+     */
+    isStoredInSecretStorage(secretStorage) {
+        let stored = true;
+        for (const name of ["master", "self_signing", "user_signing"]) {
+            stored &= secretStorage.isStored(`m.cross_signing.${name}`, false);
+        }
+        return stored;
+    }
+
+    /**
+     * Store private keys in secret storage for use by other devices. This is
+     * typically called in conjunction with the creation of new cross-signing
+     * keys.
+     *
+     * @param {SecretStorage} secretStorage The secret store using account data
+     */
+    async storeInSecretStorage(secretStorage) {
+        const getKey = this._callbacks.getCrossSigningKey;
+        for (const name of ["master", "self_signing", "user_signing"]) {
+            const encodedKey = encodeBase64(await getKey(name));
+            await secretStorage.store(`m.cross_signing.${name}`, encodedKey);
+        }
+    }
+
+    /**
+     * Get private keys from secret storage created by some other device. This
+     * also passes the private keys to the app-specific callback.
+     *
+     * @param {SecretStorage} secretStorage The secret store using account data
+     */
+    async getFromSecretStorage(secretStorage) {
+        if (!this._callbacks.saveCrossSigningKeys) {
+            throw new Error("No saveCrossSigningKeys callback supplied");
+        }
+
+        // Retrieve private keys from secret storage
+        const privateKeys = {};
+        for (const name of ["master", "self_signing", "user_signing"]) {
+            const encodedKey = await secretStorage.get(`m.cross_signing.${name}`);
+            privateKeys[name] = decodeBase64(encodedKey);
+        }
+
+        // Regenerate public keys from private keys
+        // XXX: Do we want to _also_ download public keys from the homeserver to
+        // verify they agree...?
+        // See also https://github.com/vector-im/riot-web/issues/11558
```

Comment:
Hmm - yeah, we should not be re-signing them after they're initially generated. I think you want `checkOwnCrossSigningTrust()` for this which will get the public keys from the HS (or rather, we should have already fetched them if our device list is up to date) and then try to get the matching private key from the app by calling the `getCrossSigningKey` callback. That said, looks like you're trying to move as much as possible into the js-sdk (which is probably a good thing) but that will mean we'll need to try & handle that within the js-sdk too (ie. probably call the callback first and then if that says, "no key here", look in SSSS automatically).

Reply:
Okay, I have reworked to store more things in JS SDK and rely on `checkOwnCrossSigningTrust`.